File: code/socket2/python/echo/echo_pollserv.py
import socket
import select
import sys

BUF_SIZE = 100

def error_handling(message):
    """错误处理函数"""
    print(message, file=sys.stderr)
    sys.exit(1)

def main():
    if len(sys.argv) != 2:
        print(f"Usage : {sys.argv[0]} <port>")
        sys.exit(1)
    
    try:
        # 创建TCP套接字
        serv_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    except socket.error:
        error_handling("socket() error")
    
    try:
        # 绑定地址和端口
        serv_sock.bind(('', int(sys.argv[1])))
    except socket.error:
        error_handling("bind() error")
    
    try:
        # 开始监听
        serv_sock.listen(5)
    except socket.error:
        error_handling("listen() error")
    
    # 初始化 poll 对象，并注册服务器套接字的 POLLIN 事件
    poller = select.poll()
    poller.register(serv_sock, select.POLLIN)
    
    # fd -> socket 对象的映射，poll 返回的是文件描述符
    fd_to_sock = {serv_sock.fileno(): serv_sock}
    
    while True:
        # poll 超时时间单位是毫秒，5005 毫秒 ≈ 5.005 秒
        try:
            events = poller.poll(5005)
        except select.error:
            break
        except OSError:
            break
        
        # 超时（没有任何事件就绪）
        if not events:
            continue
        
        # 遍历所有就绪的事件（fd, event）
        for fd, event in events:
            sock = fd_to_sock.get(fd)
            if sock is None:
                continue
            
            if sock == serv_sock:
                # 有新的连接请求
                try:
                    clnt_sock, clnt_addr = serv_sock.accept()
                    poller.register(clnt_sock, select.POLLIN)
                    fd_to_sock[clnt_sock.fileno()] = clnt_sock
                    print(f"connected client: {clnt_sock.fileno()} ")
                except socket.error:
                    error_handling("accept() error")
            else:
                # 有数据可读（或对端关闭/出错）
                if event & (select.POLLERR | select.POLLHUP | select.POLLNVAL):
                    # 出错或挂断，直接关闭
                    poller.unregister(fd)
                    fd_to_sock.pop(fd, None)
                    sock.close()
                    print(f"closed client: {fd} ")
                    continue
                
                if event & select.POLLIN:
                    try:
                        data = sock.recv(BUF_SIZE)
                        if not data:
                            # 客户端关闭连接
                            poller.unregister(fd)
                            fd_to_sock.pop(fd, None)
                            sock.close()
                            print(f"closed client: {fd} ")
                        else:
                            # 回声
                            sock.sendall(data)
                    except socket.error:
                        # 出错时也移除该套接字
                        try:
                            poller.unregister(fd)
                        except (KeyError, OSError):
                            pass
                        fd_to_sock.pop(fd, None)
                        sock.close()
    
    serv_sock.close()

if __name__ == "__main__":
    main()

File: code/socket2/posix/echo/echo_pollserv.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <arpa/inet.h>
#include <sys/socket.h>
#include <poll.h>

#define BUF_SIZE 100
#define MAX_CLNT 256

void error_handling(const char *buf);

int main(int argc, char *argv[])
{
    int serv_sock, clnt_sock;
    struct sockaddr_in serv_adr, clnt_adr;
    socklen_t adr_sz;
    int str_len, i;
    char buf[BUF_SIZE];

    struct pollfd poll_fds[MAX_CLNT];
    int nfds = 0;

    if (argc != 2) {
        printf("Usage : %s <port>\n", argv[0]);
        exit(1);
    }

    serv_sock = socket(PF_INET, SOCK_STREAM, 0);
    memset(&serv_adr, 0, sizeof(serv_adr));
    serv_adr.sin_family = AF_INET;
    serv_adr.sin_addr.s_addr = htonl(INADDR_ANY);
    serv_adr.sin_port = htons(atoi(argv[1]));

    if (bind(serv_sock, (struct sockaddr*)&serv_adr, sizeof(serv_adr)) == -1)
        error_handling("bind() error");
    if (listen(serv_sock, 5) == -1)
        error_handling("listen() error");

    // 监听 socket 放在数组第一个位置
    poll_fds[0].fd = serv_sock;
    poll_fds[0].events = POLLIN;
    nfds = 1;

    while (1)
    {
        int timeout_ms = 5000;  // 5 秒
        int ret = poll(poll_fds, nfds, timeout_ms);

        if (ret == -1)
            break;
        if (ret == 0)
            continue;  // 超时

        for (i = 0; i < nfds; i++)
        {
            if (poll_fds[i].revents & POLLIN)
            {
                if (poll_fds[i].fd == serv_sock)  // 新连接
                {
                    adr_sz = sizeof(clnt_adr);
                    clnt_sock = accept(serv_sock, (struct sockaddr*)&clnt_adr, &adr_sz);
                    if (clnt_sock == -1)
                        continue;

                    if (nfds < MAX_CLNT) {
                        poll_fds[nfds].fd = clnt_sock;
                        poll_fds[nfds].events = POLLIN;
                        nfds++;
                        printf("connected client: %d \n", clnt_sock);
                    } else {
                        close(clnt_sock);
                    }
                }
                else  // 客户端发来数据
                {
                    str_len = read(poll_fds[i].fd, buf, BUF_SIZE);
                    if (str_len == 0)  // 客户端关闭
                    {
                        printf("closed client: %d \n", poll_fds[i].fd);
                        close(poll_fds[i].fd);
                        // 用最后一个元素覆盖当前位置，然后 nfds--
                        poll_fds[i] = poll_fds[nfds - 1];
                        nfds--;
                        i--;  // 重新检查被搬过来的这个位置
                    }
                    else
                    {
                        write(poll_fds[i].fd, buf, str_len);  // echo
                    }
                }
            }
        }
    }

    close(serv_sock);
    return 0;
}

void error_handling(const char *buf)
{
    fputs(buf, stderr);
    fputc('\n', stderr);
    exit(1);
}
